keps/sig-etcd/4578-server-feature-gate/README.md

```diff
@@ -0,0 +1,364 @@
+
+```go
+defaultServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+```
+
+The feature gates for a server can only be set with the `--feature-gates`flag or in the `config-file` during startup. We do not support dynamically changing the feature gates when the server is running.
+
+The `ServerConfig` struct will have a new `featureGate k8s.io/component-base/featuregate.FeatureGate`(immutable) field. `EtcdServer` would have an interface of `FeatureEnabled(key Feature) bool`, and it can be referenced throughout the server.
+
+```go
+type ServerConfig struct {
+  ...
+  // ServerFeatureGate server level feature gate
+  ServerFeatureGate featuregate.FeatureGate
+  ...
+}
+
+func (s *EtcdServer) FeatureEnabled(key Feature) bool
+```
+
+### Feature Stages
+
+Following the convention of Kubernetes, a feature can go through a lifecycle of Alpha → Beta → GA → Deprecated.
+
+| Feature Stage | Properties | Graduation Criteria |
+| --- | --- | --- |
+| Alpha | <ul><li>Disabled by default. </li><li>Might be buggy. Enabling the feature may expose bugs. </li><li>Support for feature may be dropped at any time without notice. </li><li>The API may change in incompatible ways in a later software release without notice. </li><li>Recommended for use only in short-lived testing clusters, due to increased risk of bugs and lack of long-term support.</li></ul> | Before moving a feature to Beta, it should have <ul><li> Full unit/integration/e2e/robustness test coverage.</li><li>Full performance benchmark/test if applicable.</li><li> No significant changes for at least 1 minor release.</li><li>Other feature specific criteria in the feature KEP. </li></ul> |
+| Beta | <ul><li>Enabled by default. </li><li>The feature is well tested. Enabling the feature is considered safe.</li><li>Support for the overall feature will not be dropped, though details may change.</li><li>Recommended for only non-business-critical uses because of potential for discovering new hard-to-spot bugs through wider adoption.</li></ul> | Before moving a feature to GA, it should have <ul><li> Widespread usage.</li><li>No bug reported for at least 1 minor release.</li><li>Other feature specific criteria in the feature KEP. </li></ul> |
+| GA | <ul><li>The feature is always enabled; you cannot disable it.</li><li>The corresponding feature gate is no longer needed.</li><li>Stable versions of features will appear in released software for many subsequent versions.</li></ul> | Before deprecating a GA feature, it should have <ul><li> Feature deprecation announcement.</li><li>No user impacting change for at least 1 minor release.</li><li>Other feature specific criteria in the feature KEP. </li></ul> |
```

Comment:
> GA | The feature is always enabled; you cannot disable it.

Question: does it mean that a GAed feature will never have a bool flag (i.e. `--pre-vote`) for users to enable or disable it?

When a feature is GAed, it only means that its quality meets some criteria. But it may still make sense for users to enable or disable it for different scenarios or purpose, and it's exactly the reason why there are still some bool flags in api-server and etcd.

Reply:
feature gates are not replacement of boolean flags. They are different concepts. Feature gate is meant to facilitate the development of a feature. A feature could have many other flags, including boolean flags. Here is one example in K8s: https://github.com/kubernetes/kubernetes/blob/5a121aad53a819c3c9b26a6b5866b1ea8c686f5b/pkg/kubelet/apis/config/validation/validation.go#L142

Reply:
We should utilize feature flags for all new features to clearly indicate their maturity to users. While most features should be enabled by default at GA, some features like pre-vote introduce tradeoffs that users might want to control. In these cases, a separate toggle could be provided.

The beta stage is the ideal time to assess whether a feature should be permanently enabled by default or if an off switch is warranted. We need to ensure the feature consistently provides positive value with minimal downsides users might want to mitigate. For GA, the feature flag would be locked to its default state, with the potential addition of a separate boolean flag if needed.